File: C++/Graphs/AdjacencyField.cpp
/*
Explanation of Approach:
- We represent the graph as an Adjacency List, which is an array of vectors.
- Each vector in the array represents a vertex and contains a list of its adjacent vertices.
- We use an undirected graph representation, so for every edge (u, v), we add both (u -> v) and (v -> u) to the adjacency list.

Time Complexity:
- Adding an edge takes O(1) time since it involves only inserting an element into the vector.
- Printing the adjacency list takes O(V + E) time, where V is the number of vertices and E is the number of edges, as we need to traverse through the entire adjacency list.

Space Complexity:
- The space complexity of the adjacency list is O(V + E), where V is the number of vertices and E is the number of edges, as we store all the edges in the adjacency list.

Sample Input:
```
Enter the number of vertices: 5
Enter the number of edges: 4
Enter edges (u, v):
0 1
0 2
1 3
3 4
```

Sample Output:
Adjacency List:
Adjacency list of vertex 0: 1 2 
Adjacency list of vertex 1: 0 3 
Adjacency list of vertex 2: 0 
Adjacency list of vertex 3: 1 4 
Adjacency list of vertex 4: 3 

In the sample input, we created a graph with 5 vertices and 4 edges, and the output shows the adjacency list representation of the graph.
*/


#include <iostream>
#include <vector>

using namespace std;

// Function to add an edge to the adjacency list
void addEdge(vector<vector<int>>& adjList, int u, int v) {
    adjList[u].push_back(v);
    adjList[v].push_back(u); // For undirected graph, add both edges
}

// Function to print the adjacency list
void printAdjacencyList(const vector<vector<int>>& adjList) {
    int V = adjList.size();
    for (int i = 0; i < V; ++i) {
        cout << "Adjacency list of vertex " << i << ": ";
        for (int v : adjList[i]) {
            cout << v << " ";
        }
        cout << endl;
    }
}

int main() {
    int V, E; // Number of vertices and edges
    cout << "Enter the number of vertices: ";
    cin >> V;
    cout << "Enter the number of edges: ";
    cin >> E;

    // Create an adjacency list of size V
    vector<vector<int>> adjList(V);

    cout << "Enter edges (u, v):" << endl;
    for (int i = 0; i < E; ++i) {
        int u, v;
        cin >> u >> v;
        addEdge(adjList, u, v);
    }

    // Print the adjacency list
    cout << "Adjacency List:" << endl;
    printAdjacencyList(adjList);

    return 0;
}

File: Python/Graphs/AdjacencyField.py
'''
Explanation of Approach:
- We represent the graph using an adjacency list, which is a dictionary where each vertex is a key, and the value for each key is a list of vertices that are adjacent to it.
- The `create_empty_graph` function initializes an empty dictionary to represent the graph.
- The `add_edge` function takes the graph, vertex1, and vertex2 as input and adds an edge between them. It first checks if the vertices exist in the graph. If not, it adds them as keys to the dictionary with an empty list as their value. Then, it appends each vertex to the other's adjacency list to represent the undirected nature of the edge.
- The `sample_graph` function provides a sample input by creating a graph with five vertices (1, 2, 3, 4) and adding edges between them.
- The sample output is the adjacency list representation of the sample graph.

Time Complexity:
- Adding an edge using the `add_edge` function takes O(1) time because dictionary operations (adding elements to a list) and list appends take constant time on average.

Space Complexity:
- The space complexity of the graph representation using an adjacency list is O(V + E), where V is the number of vertices and E is the number of edges. In the worst case, it can be O(V^2) for dense graphs.

Sample Input:
graph = sample_graph()

Sample Output:
{1: [2, 3], 2: [1, 3, 4], 3: [1, 2, 4], 4: [2, 3]}

This output represents a graph with four vertices (1, 2, 3, 4) and five edges connecting them as shown in the following structure:
     1
   /   \
  2-----3
   \   /
     4
'''

# Define a function to create an empty graph with no vertices
def create_empty_graph():
    return {}

# Define a function to add an edge between two vertices in the graph
def add_edge(graph, vertex1, vertex2):
    if vertex1 not in graph:
        graph[vertex1] = []
    if vertex2 not in graph:
        graph[vertex2] = []

    graph[vertex1].append(vertex2)
    graph[vertex2].append(vertex1)

# Sample input to create a graph
def sample_graph():
    graph = create_empty_graph()
    add_edge(graph, 1, 2)
    add_edge(graph, 1, 3)
    add_edge(graph, 2, 3)
    add_edge(graph, 2, 4)
    add_edge(graph, 3, 4)
    return graph

# Sample output for the above sample input:
# {1: [2, 3], 2: [1, 3, 4], 3: [1, 2, 4], 4: [2, 3]}

# Uncomment the following lines to test the sample input and output
# graph = sample_graph()
# print(graph)

